design/baremetal-operator/bulk-set-bios-config.md

```diff
@@ -0,0 +1,218 @@
+
+The BIOS Settings are retrieved from the BMC by Ironic and cached
+whenever the node moves to `manageable` or `cleaning`, or when the settings
+are updated. Recent changes to Ironic (see the Ironic proposal
+[BIOS Registry](https://review.opendev.org/c/openstack/ironic-specs/+/774681)
+added an additional query of the BIOS Registry.
+
+The baremetal-operator can get the BIOS Settings from Ironic after the node
+has been moved to manageable via the `v1/nodes/{node_ident}/bios?detail=True`
+API.  This API provides a list of BIOS Settings (aka `Attributes`) in JSON and each
+setting includes the following data:
+
+- Name
+- Value
+- Registry information
+  - Description
+  - Type (e.g. enum, string, integer, or boolean)
+  - ReadOnly (a boolean indicating whether the attribute is read only)
+  - Immutable (a boolean indicating whether or not the attribute is changeable)
+  - Help text (localized)
+  - Allowable values for enum type attributes
+  - Minimum and maximum values (for numeric attributes)
+  - Minimum and maximum character lengths (for string attributes)
+
+Upon retrieving the BIOS Settigs from Ironic, the baremetal-operator will
+populate the 'currentBIOSSettings` field for each host, represented as a
+ConfigMap with an annotation in BareMetalHostStatus. As there are no fields
+in BIOS configuration that are secret (i.e. no passwords etc.) this new entry
+does not need to be secret-protected. 
+
+From `currentBIOSSettings` the BMO will also generate the
+`updateBIOSSettingsTemplate` which is a list consisting only of `Name`/`Value`
+pairs per Setting. The BMO will use the `ReadOnly` field in the Registry
+information to build this list. Settings with `ReadOnly` set to `True` will be
+ignored to avoid adding these settings via Ironic as they can't be written to
+a BMC.  'updateBIOSettingsTemplate' will be stored in the CRD, referenced in
+the BareMetalHostStatus.
+
+A user can create 'updateBIOSSettings' by copying the contents of
+'updateBIOSSettingsTemplate' to it and updating any values that need changing.
+Likewise, 'updateBIOSSettingsTemplate' can be copied to 'updateBIOSSettings'
+on multiple hosts of the same type to effectively clone the settings from a
+reference host, ensuring that all similar systems have the same BIOS
+configuration.
+
+The baremetal-operator will detect when 'updateBIOSSettings' changes by
+comparing the 'Name'/'Value' pairs in 'updateBIOSSettings' to the current BIOS
+configuration in 'currentBIOSSettings'. When a change is detected, the BMO will
+add the new values to the Ironic clean-steps API. Before adding it though, the
+BMO will do validation checks on the new values using the Registry information
+in 'currentBIOSSettings'. For example, integer types will checked against
+minimum and maximum parameters and enumeration types will be compared against
+allowable values. If the validation fails a message will be logged and the
+BIOS Setting will not be added to clean-steps.
```

Comment:
Arda is in the process of adding a validating webhook, maybe that would be the place to verify the data so that the user is prevented from setting invalid values, rather than just dropping them on the floor.

Reply:
OK great, I added a note that we can leverage the validating webhook if possible.